File: solutions/haskell/rational-numbers/1/src/RationalNumbers.hs
module RationalNumbers
(Rational,
 abs,
 numerator,
 denominator,
 add,
 sub,
 mul,
 div,
 pow,
 expRational,
 expReal,
 rational) where

import Prelude hiding (div, abs, Rational)

-- Data definition -------------------------------------------------------------
data Rational a = R a a deriving(Eq, Show)

rational :: Integral a => (a, a) -> Rational a
rational (a,b) = R (quot a' m) (quot b' m)
  where (a',b') = if b < 0 then (-a, -b) else (a, b)
        m = gcd a b

-- unary operators -------------------------------------------------------------
abs :: Integral a => Rational a -> Rational a
abs (R a b) = rational ((if a < 0 then (-a) else a),(if b < 0 then (-b) else b))

numerator :: Integral a => Rational a -> a
numerator (R a _) = a

denominator :: Integral a => Rational a -> a
denominator (R _ b) = b

-- binary operators ------------------------------------------------------------
add :: Integral a => Rational a -> Rational a -> Rational a
add (R a1 b1) (R a2 b2) = rational ((a1 * b2 + a2 * b1),(b1 * b2))

sub :: Integral a => Rational a -> Rational a -> Rational a
sub (R a1 b1) (R a2 b2) = rational ((a1 * b2 - a2 * b1),(b1 * b2))

mul :: Integral a => Rational a -> Rational a -> Rational a
mul (R a1 b1) (R a2 b2) = rational ((a1 * a2),(b1 * b2))

div :: Integral a => Rational a -> Rational a -> Rational a
div (R a1 b1) (R a2 b2) = rational ((a1 * b2),(a2 * b1))

pow :: Integral a => Rational a -> a -> Rational a
pow (R a b) n = if n >= 0 
                then rational ((a ^ n),(b ^ n))
                else rational ((b ^ (-n)),(a ^ (-n)))

expRational :: (Integral a, Floating b) => Rational a -> b -> b
expRational (R a b) n = fromIntegral a ** n / fromIntegral b ** n

expReal :: Floating a => Integral b => a -> Rational b -> a
expReal n (R a b) = (n ^^ fromIntegral a) ** (1 / fromIntegral b)

File: solutions/c/rational-numbers/1/rational_numbers.c
#include "rational_numbers.h"
#include <math.h>
#include <stdlib.h>

rational_t add(rational_t a, rational_t b) {
  rational_t res = {
      .numerator = a.numerator * b.denominator + b.numerator * a.denominator,
      .denominator = a.denominator * b.denominator,
  };
  return reduce(res);
}

rational_t subtract(rational_t a, rational_t b) {
  rational_t res = {
      .numerator = a.numerator * b.denominator - b.numerator * a.denominator,
      .denominator = a.denominator * b.denominator,
  };
  return reduce(res);
}

rational_t multiply(rational_t a, rational_t b) {
  rational_t res = {
      .numerator = a.numerator * b.numerator,
      .denominator = a.denominator * b.denominator,
  };
  return reduce(res);
}

rational_t divide(rational_t a, rational_t b) {
  rational_t res = {
      .numerator = a.numerator * b.denominator,
      .denominator = b.numerator * a.denominator,
  };
  return reduce(res);
}

rational_t absolute(rational_t a) {
  rational_t res = {
      .numerator = abs(a.numerator),
      .denominator = abs(a.denominator),
  };
  return reduce(res);
}

rational_t exp_rational(rational_t r, double n) {
  rational_t res;
  if (round(n) == n) {
    if (n >= 0) {
      res.numerator = pow(r.numerator, n);
      res.denominator = pow(r.denominator, n);
    } else {
      double m = -n;
      res.numerator = pow(r.denominator, m);
      res.denominator = pow(r.numerator, m);
    }
  } else {
    res.numerator = pow(r.numerator, n);
    res.denominator = pow(r.denominator, n);
  }
  return reduce(res);
}

double exp_real(double x, rational_t r) {
  return pow(x, (double)r.numerator / (double)r.denominator);
}

static int gcd(int a, int b) {
  if (a == 0) {
    return b;
  }
  if (b == 0) {
    return a;
  }
  return gcd(b, a % b);
}

rational_t reduce(rational_t r) {
  int c;
  if (abs(r.numerator) > abs(r.denominator)) {
    c = gcd(abs(r.numerator), abs(r.denominator));
  } else {
    c = gcd(abs(r.denominator), abs(r.numerator));
  }
  rational_t res = {
      .numerator = r.numerator / c,
      .denominator = r.denominator / c,
  };
  if (res.numerator < 0 && res.denominator < 0) {
    res.numerator = -res.numerator;
    res.denominator = -res.denominator;
  } else if (res.denominator < 0) {
    res.numerator = -res.numerator;
    res.denominator = -res.denominator;
  }
  return res;
}
